File: Leetcode/dfs_recursion_backtracking/200_NumberOfIslands/Solution.java
class Solution {
    public int numIslands(char[][] grid) {
        if(grid==null||grid.length==0)
            return 0;
        int res = 0;
        for(int i = 0; i<grid.length; ++i){
            for(int j = 0; j<grid[i].length; ++j){
                if(grid[i][j]=='0')
                    continue;
                dfsSink(grid, i, j);
                res += 1;
            }
        }
        return res;
    }
        private void dfsSink(char[][] grid,
                             int i, int j){
            if(i<0||j<0||i>=grid.length||j>=grid[0].length||grid[i][j]=='0')
                return;
            grid[i][j] = '0';
            dfsSink(grid, i+1, j);dfsSink(grid, i-1, j);
            dfsSink(grid, i, j+1);dfsSink(grid, i, j-1);
        }
}

File: Leetcode/dfs_recursion_backtracking/200_NumberOfIslands/Solution.py
class Solution(object):
    def numIslands(self, grid):
        """
        :type grid: List[List[str]]
        :rtype: int
        """
        res = 0
        if not grid: return res
        def dfs(i, j):
            if not(0<=i<len(grid) and 0<=j<len(grid[i]) and grid[i][j]=='1'):
                return
            grid[i][j] = '0'
            dfs(i+1, j); dfs(i-1, j); dfs(i, j+1); dfs(i, j-1);
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j]=='0':
                    continue
                dfs(i, j)
                res += 1
        return res
